### PR TITLE
feat(data): support SSR response headers and cookies

### PR DESCRIPTION
**Pull request by kojiwakayama**
## Summary

- add typed `headers` and `cookies` response metadata to `getServerData`
- preserve metadata across direct and isolated worker execution, layout and page merging, normal SSR, redirects, and 404 responses
- document the public contract and regenerate API references

## Contract and safety

- layouts merge outermost to innermost, then the page; the closest loader wins duplicate custom headers and cookies remain distinct
- framework-owned CORS, cache, content, security, transport, and `x-veryfront-*` headers are rejected
- cookie responses force `no-cache`, suppress ETags, and are not persisted in the render cache
- `getStaticData` rejects response metadata, including thrown control results, so static caches cannot replay cookies
- cookie values never enter enumerable error context, logs, or telemetry
- response metadata applies to full document responses; API routes or middleware remain the write surface for client-side page-data navigation

## Red-green TDD

The initial focused tests failed at the fetcher, render pipeline, SSR service, response builder, redirect, and 404 seams because metadata was dropped. Additional red tests exposed and fixed two security edges: thrown static redirects bypassing the cache guard, and cookie values entering enumerable error context.

## Verification

- focused Deno tests: 16 suites, 380 steps, all passed
- `deno task typecheck`
- `deno task fmt:check`
- `deno task lint`
- `deno task lint:client-bundle`
- `deno task docs:api-reference:check`
- `deno task docs:validate`
- dependency, module, style, barrel JSDoc, and anti-slop checks

No UI changes.

## Summary by CodeRabbit

* **New Features**
  * Server-rendered pages can return custom response headers and cookies.
  * Redirects and not-found responses preserve supplied metadata.
  * Added cookie attributes, including expiration, security, and SameSite settings.
  * Added public types for response metadata, cookies, data results, and static data results.
* **Bug Fixes**
  * Improved validation and protection against invalid or unsafe response metadata.
  * Responses containing cookies are no longer incorrectly cached.
* **Documentation**
  * Expanded guidance and API references for headers, cookies, caching, and redirects.

### CODE REVIEW

**Comment by kojiwakayama** [comment]
@codex review

File: src/server/handlers/request/ssr/ssr.handler.ts [hunk 1]
```diff
@@ -373,15 +377,21 @@ export class SSRHandler extends BaseHandler {
...
 
     const customResponse = await tryErrorPageFallback(req, ctx, builder, {
       statusCode: 404,
       pathname: slug || "/",
     }, dependencySnapshot);
-    if (customResponse) return this.respond(customResponse);
+    if (customResponse) {
+      appendDataResponseMetadata(customResponse.headers, result);
+      return this.respond(customResponse);
+    }
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/server/handlers/request/ssr/ssr.handler.ts:392]: Fixed in 93449c71e. The custom runtime/server-error fallback now appends normalized response metadata before responding. A focused red-green handler test covers both the custom header and Set-Cookie field.

File: src/data/schemas/data.schema.ts [hunk 2]
```diff
@@ -53,20 +79,36 @@ export const getCacheEntrySchema = defineSchema((v) =>
...
+/** Props, routing control, caching, and response metadata returned from `getServerData()`. */
 export type DataResult<T = unknown> = InferSchema<ReturnType<typeof getDataResultSchema>> & {
   props?: T;
 };
+/** Cache-safe result returned from `getStaticData()`. */
+export type StaticDataResult<T = unknown> =
+  & InferSchema<
+    ReturnType<typeof getStaticDataResultSchema>
+  >
+  & { props?: T };
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/schemas/data.schema.ts]: Fixed in 93449c71e. StaticDataResult now declares headers?: never and cookies?: never. A compile-time PageWithData fixture rejects a metadata-bearing DataResult, and the test-typecheck baseline passes with 0 new failures.

File: src/rendering/orchestrator/pipeline.behavior.test.ts [hunk 3]
```diff
@@ -27,6 +27,7 @@ import {
   globalModuleCache,
 } from "#veryfront/modules/react-loader/ssr-module-loader/cache/index.ts";
 import { hashString } from "#veryfront/cache/hash.ts";
+import { resolveSSRControlOutcome } from "../ssr-outcome.ts";
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/orchestrator/pipeline.behavior.test.ts]: Fixed in 93449c71e. The cross-directory test import now uses #veryfront/rendering/ssr-outcome.ts. Dependency and module boundary checks pass.

File: src/rendering/orchestrator/render-result-assembly.ts [hunk 4]
```diff
@@ -49,10 +52,12 @@ export function assembleRenderResult(options: AssembleRenderResultOptions): Rend
     nodeMap: options.pageBundle.nodeMap,
     stream: options.ssrResult.finalStream,
     ssrHash: options.ssrResult.ssrHash,
+    ...(options.headers ? { headers: options.headers } : {}),
+    ...(options.cookies ? { cookies: options.cookies } : {}),
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/orchestrator/render-result-assembly.ts]: Fixed in 93449c71e. Renderer now carries headers and cookies through CachedRenderData and reconstruction. Focused red-green coverage also verifies cookie responses bypass the outer render cache and safe custom headers survive cache hits.

**Comment by kojiwakayama** [comment]
Addressed all four review threads in exact head `93449c71e`:

- Renderer carries headers/cookies through singleflight reconstruction.
- Cookie-bearing renders bypass the outer render cache; cacheable custom headers survive cache hits.
- `StaticDataResult` forbids `headers` and `cookies` at compile time.
- The cross-directory import uses the internal alias.
- Custom runtime/server-error pages retain headers and distinct `Set-Cookie` fields.

Red-green evidence: new tests first failed at each missing boundary, then passed after the fixes. The affected surface is green at 22 suites/files and 455 steps. Full Renderer/cache/handler coverage is green at 126 and 53 steps respectively. Typecheck, format, lint, test-typecheck, API reference generation, docs validation (51 tests, 1,383 links), client-bundle, dependency-boundary, module-boundary, barrel-JSDoc, anti-slop, and style checks all pass.

The CodeRabbit docstring percentage is a non-blocking heuristic for private implementation and test helpers. The repository public documentation and JSDoc gates pass, and generated API references are current.

All four review threads are replied to and resolved. Replacement GitHub CI is running.

@codex review

File: src/rendering/renderer.ts [hunk 5]
```diff
@@ -866,6 +868,8 @@ export class Renderer {
       headings: cachedData.headings,
       ssrHash: cachedData.ssrHash,
       pageModule: cachedData.pageModule,
+      ...(cachedData.headers ? { headers: cachedData.headers } : {}),
+      ...(cachedData.cookies ? { cookies: cachedData.cookies } : {}),
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/renderer.ts:921]: Fixed in 4f63ef633. A follower that observes a cookie-bearing shared result now performs its own uncached render with its own request and abort signal. The new overlapping-request regression proves leader and follower receive distinct cookie values and no cookie-bearing result is persisted.

[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/renderer.ts:921]: Fixed in 4f63ef633. This is the duplicate credential-replay report: cookie-bearing singleflight followers rerender independently, and the regression asserts request-specific leader/follower cookies plus zero cache persistence.

File: src/rendering/shared/context-aware-cache.ts [hunk 6]
```diff
@@ -358,6 +358,9 @@ export class ContextAwareCacheCoordinator {
     if (result.pageModule) {
       cloned.pageModule = { ...result.pageModule };
     }
+    if (result.headers) {
+      cloned.headers = { ...result.headers };
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/shared/context-aware-cache.ts:362]: Fixed in 4f63ef633. The cache snapshot and serialized codec now retain custom response headers after revalidating and detaching them. The focused round-trip test covers both in-memory and serialized payloads.

File: src/rendering/orchestrator/pipeline.ts [hunk 7]
```diff
@@ -971,6 +1015,8 @@ export class RenderPipeline {
                 cacheCoordinator: this.config.cacheCoordinator,
                 logger: renderPipelineLog,
                 nonce: renderOptions.nonce,
+                headers: responseHeaders,
+                cookies: responseCookies,
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/orchestrator/pipeline.ts:1039]: Fixed in 4f63ef633. The pipeline keeps accumulated headers and cookies visible to the outer error path and attaches their merged metadata through the existing WeakMap carrier before rethrowing. The regression covers an SSR failure after successful server-data resolution.

**Comment by kojiwakayama** [comment]
Pushed `4f63ef633` and addressed all four exact-head review threads.

Red-green coverage:
- Concurrent singleflight leader/follower requests originally replayed the leader cookie. The regression now proves each request receives its own cookie and cookie-bearing results are not persisted.
- Memory and serialized cache payloads originally dropped custom response headers. Both codecs now preserve validated, detached headers.
- An SSR failure after successful server-data resolution originally lost collected headers and cookies. The error path now carries the merged metadata through the non-enumerable WeakMap attachment.

Verification is green:
- focused rendering tests: 7 files, 115 steps
- broader cache/store/SSR tests: 16 files, 303 steps
- `deno check`
- `deno task typecheck`
- `deno task lint`
- `deno task lint:test-typecheck` (0 new findings)
- `deno task fmt:check`
- dependency and module boundaries
- anti-slop and style gates
- generated API reference check
- docs validation: 51 tests, 106 steps, 1,383 links
- `git diff --check`

All reported threads are replied to and resolved. Please review the new exact head.

@codex review

**Comment by kojiwakayama** [comment]
Replacement head `fc302ff9b` fixes the integration regression exposed by CI.

The new metadata validator had accidentally rejected two established data-result behaviors that are outside this feature's scope:

- redirect takes precedence over not-found and props, and not-found takes precedence over props;
- finite negative `revalidate` values remain accepted.

The fix restores those semantics in both direct execution and the isolated worker boundary while continuing to validate and retain the new response headers and cookies.

Red-green evidence:
- RED: focused validator regression failed both precedence and negative revalidation cases, matching CI's four failures.
- GREEN: the original failing integration file passes all 47 steps.
- GREEN: validator, response metadata, server/static fetchers, integration edge cases, and isolated worker coverage pass 200 steps.
- Formatting, lint, typecheck, test-typecheck, and `git diff --check` pass.

Replacement CI is running. Please review the new exact head.

@codex review

File: src/data/types.ts [hunk 8]
```diff
@@ -18,7 +31,7 @@ export interface PageWithData<T = unknown> {
   getServerData?: (context: DataContext) => DataResult<T> | Promise<DataResult<T>>;
   getStaticData?: (
     context: Omit<DataContext, "request" | "query">,
-  ) => DataResult<T> | Promise<DataResult<T>>;
+  ) => StaticDataResult<T> | Promise<StaticDataResult<T>>;
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/types.ts]: Fixed in 89634fa80. Metadata-free redirect() and notFound() overloads return StaticDataResult, while calls that supply response metadata retain DataResult. The compile-time PageWithData regression now passes and metadata-bearing static results remain rejected.

File: src/data/data-result-validation.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      (redirectRecord.permanent !== undefined && typeof redirectRecord.permanent !== "boolean")
+    ) return fail();
+    redirectDestination = redirectRecord.destination;
+    redirectPermanent = redirectRecord.permanent as boolean | undefined;
+  }
+  if (notFound !== undefined && typeof notFound !== "boolean") return fail();
+  if (
+    revalidate !== undefined && revalidate !== false &&
+    (typeof revalidate !== "number" || !Number.isFinite(revalidate))
+  ) return fail();
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/data-result-validation.ts]: Fixed in 89634fa80. Revalidation is now validated only after redirect and not-found precedence, in both direct validation and the isolated worker snapshot. The red tests used Infinity and a string in discarded revalidate fields; direct and worker regressions are green.

File: src/rendering/orchestrator/pipeline.ts [hunk 10]
```diff
@@ -980,6 +1027,18 @@ export class RenderPipeline {
             } catch (error) {
               if (error instanceof Error) {
                 (error as Error & { sourceFile?: string }).sourceFile = sourceFile;
+                if (responseHeaders || responseCookies) {
+                  attachDataResponseMetadata(
+                    error,
+                    mergeDataResponseMetadata([
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/orchestrator/pipeline.ts]: Fixed in 89634fa80. SSRService now reads the non-enumerable attachment when constructing runtime, server-error, and app-router-error-boundary results. Focused red-green coverage verifies headers and cookies survive all three non-control failure paths.

**Comment by kojiwakayama** [comment]
Pushed exact head 89634fa80 and addressed all three newly reported findings.

Red-green coverage now proves:
- attached headers and cookies reach runtime, server-error, and app-router error-boundary responses;
- metadata-free redirect() and notFound() remain assignable from getStaticData;
- discarded revalidate fields do not override redirect or not-found precedence in direct or isolated execution.

Verification is green: 29 affected suites, 530 steps, focused compile checks, formatting, lint, typecheck, test-typecheck with 0 new findings, generated API reference validation, and git diff checks. All review threads are replied to and resolved. Replacement CI is running.

@codex review

File: src/data/response-metadata.ts [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+
+/**
+ * Carry validated response metadata through an internal error path without
+ * placing cookie values in enumerable error context, logs, or telemetry.
+ */
+export function attachDataResponseMetadata<T extends Error>(
+  carrier: T,
+  metadata: DataResponseMetadata,
+): T {
+  attachedResponseMetadata.set(carrier, normalizeDataResponseMetadata(metadata));
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/response-metadata.ts:322]: Fixed in 5132378e0. Post-data failures now carry metadata on a fresh request-local Error wrapper instead of mutating the project Error. SSRService unwraps the original error for redirect, app-router, reporting, and stack classification while applying metadata from the unique carrier. The red regression throws one module-scoped Error twice; only the first render now receives its cookie.

**Comment by kojiwakayama** [comment]
Pushed exact head 5132378e0 for the reusable-Error isolation finding.

Red-green coverage rethrows one module-scoped Error across two requests. Before the fix, the second response inherited the first request session cookie. After the fix, post-data metadata lives on a unique request-local carrier, while the original error remains authoritative for redirect, app-router, reporting, and stack behavior.

Broader verification is green at 40 suites and 723 steps across data, renderer, cache, pipeline, worker, SSR handler, service, and integration coverage. Formatting, lint, typecheck, test-typecheck with 0 new findings, API reference validation, and diff checks pass. The review thread is replied to and resolved. Replacement CI is running.

@codex review

File: src/data/data-result-validation.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    normalized.redirect = {
+      destination: redirectDestination,
+      ...(redirectPermanent !== undefined ? { permanent: redirectPermanent } : {}),
+    };
+  } else if (notFound === true) {
+    normalized.notFound = true;
+  } else {
+    if (
+      revalidate !== undefined && revalidate !== false &&
+      (typeof revalidate !== "number" || !Number.isFinite(revalidate))
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/data-result-validation.ts]: Fixed in 3fe7a5d01. Direct data-result validation and the isolated worker boundary now reject finite negative revalidation values while still applying redirect/not-found precedence before validating discarded fields. The red-green regressions cover both execution paths.

File: src/server/services/rendering/ssr.service.ts [hunk 13]
```diff
@@ -303,8 +320,15 @@ export class SSRService implements SSRServiceLike {
       }
 
       const isStreaming = !!result.stream && !result.html;
-      const cacheStrategy = useNoCache ? "no-cache" : "short";
-      const etag = isStreaming ? undefined : computeSSRETag(result.ssrHash, result.html);
+      const responseMetadata: DataResponseMetadata = {
+        ...(result.headers ? { headers: result.headers } : {}),
+        ...(result.cookies ? { cookies: result.cookies } : {}),
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/server/services/rendering/ssr.service.ts:325]: Fixed in 3fe7a5d01. Late streaming redirect and not-found controls now inherit the response metadata already resolved for that request. The merge keeps prior cookies once, appends control cookies, and lets the later control override duplicate custom headers. Red-green coverage exercises both late control outcomes.

**Comment by kojiwakayama** [comment]
Exact-head update `3fe7a5d01`

- Rejects negative revalidation intervals in direct and isolated-worker validation while preserving control-result precedence.
- Preserves resolved headers and cookies when a streaming render reports a late redirect or not-found control.
- Verifies merge precedence and ensures inherited cookies are not duplicated.
- Red tests reproduced both review findings before implementation.
- Expanded affected-area run: 130 suites, 2,189 steps, all green.
- `deno check`, format, lint, test type-check baseline, full typecheck, API reference check, and `git diff --check` pass.

@codex review

File: src/data/schemas/data.schema.ts [hunk 14]
```diff
@@ -18,8 +18,34 @@ export const getRedirectSchema = defineSchema((v) =>
...
+  })
+);
+
 /** Result returned from data fetching functions */
 export const getDataResultSchema = defineSchema((v) =>
+  v.object({
+    props: v.unknown().optional(),
+    redirect: getRedirectSchema().optional(),
+    notFound: v.boolean().optional(),
+    revalidate: v.union([v.number(), v.literal(false)]).optional(),
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/schemas/data.schema.ts]: Fixed in 67d2b4cc9. Both exported schemas now require a non-negative revalidate number, matching validateDataResult and isolated-worker validation. Added direct schema regressions for -1 and 0, and updated the integration contract that previously expected negative values to pass.

File: src/rendering/orchestrator/pipeline.ts [hunk 15]
```diff
@@ -635,23 +652,43 @@ export class RenderPipeline {
     dataResults: FetchedDataResult[],
     pageProps: Record<string, unknown>,
     layoutProps: Map<string, Record<string, unknown>>,
-  ): void {
-    for (const { type, id, result, error } of dataResults) {
+  ): DataResponseMetadata {
+    for (const { error } of dataResults) {
       if (error) throw error;
+    }
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/orchestrator/pipeline.ts]: Fixed in 67d2b4cc9. Successful page/layout results are merged before a parallel data failure is rethrown. When metadata exists, the original failure is carried by a fresh request-local Error with the merged metadata. The regression runs a successful layout beside a failing page and asserts both the original error identity and layout header/cookie.

File: src/rendering/orchestrator/pipeline.ts [hunk 16]
```diff
@@ -980,6 +1027,17 @@ export class RenderPipeline {
             } catch (error) {
               if (error instanceof Error) {
                 (error as Error & { sourceFile?: string }).sourceFile = sourceFile;
+                if (responseHeaders || responseCookies) {
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/orchestrator/pipeline.ts]: Fixed in 67d2b4cc9. The outer render catch now wraps arbitrary thrown values when data metadata exists. The request-local carrier retains the original non-Error control in a WeakMap, uses a fixed non-sensitive message, and SSRService unwraps it before classification. Regressions cover pipeline and end-to-end 404 metadata behavior, plus an undefined-failure edge case.

**Comment by kojiwakayama** [comment]
Addressed all three findings in `67d2b4cc9` with RED/GREEN regressions.

Verification:
- focused metadata, schema, pipeline, SSR, and integration tests: green
- affected sweep: 130 suites, 2,193 steps, green
- `deno task fmt:check`: green
- `deno task lint`: green
- `deno task typecheck`: green
- `deno task lint:test-typecheck`: green
- `deno task docs:api-reference:check`: green
- `git diff --check`: green

All review threads are resolved. @codex review

File: src/data/schemas/data.schema.ts [hunk 17]
```diff
@@ -18,13 +18,39 @@ export const getRedirectSchema = defineSchema((v) =>
...
+    cookies: v.array(getResponseCookieSchema()).optional(),
+  })
+);
+
+/** Cache-safe result returned from getStaticData. */
+export const getStaticDataResultSchema = defineSchema((v) =>
+  v.object({
+    props: v.unknown().optional(),
+    redirect: getRedirectSchema().optional(),
+    notFound: v.boolean().optional(),
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/schemas/data.schema.ts:84]: Fixed in `8e368cf08`. `StaticDataResultSchema` now declares `headers` and `cookies` as always-invalid optional fields, so the schema rejects either field instead of stripping it. The regression test covers both cases in Deno and Bun.

File: src/data/schemas/data.schema.ts [hunk 18]
```diff
@@ -18,13 +18,39 @@ export const getRedirectSchema = defineSchema((v) =>
...
 /** Result returned from data fetching functions */
 export const getDataResultSchema = defineSchema((v) =>
   v.object({
     props: v.unknown().optional(),
     redirect: getRedirectSchema().optional(),
     notFound: v.boolean().optional(),
-    revalidate: v.union([v.number(), v.literal(false)]).optional(),
+    revalidate: v.union([v.number().nonnegative(), v.literal(false)]).optional(),
+    headers: v.record(v.string(), v.string()).optional(),
+    cookies: v.array(getResponseCookieSchema()).optional(),
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/schemas/data.schema.ts:73]: Fixed in `8e368cf08`. `DataResultSchema` now refines parsed metadata through `normalizeDataResponseMetadata`, the same validator used by runtime execution. The contract tests cover a framework-owned header, fractional `maxAge`, unknown cookie fields, and a valid metadata result.

**Comment by kojiwakayama** [comment]
Exact head `8e368cf0801751b1727a05e1b22d5b2b2044193e` addresses both latest schema findings. RED/GREEN coverage now proves static response metadata is rejected and exported dynamic-data schemas enforce the same header and cookie rules as runtime normalization. Local affected sweep: 130 suites, 2,195 steps. Formatting, lint, typecheck, test-typecheck baseline, generated API reference check, Bun-focused tests, and diff check are green. @codex review

File: src/server/services/rendering/ssr.service.ts [hunk 19]
```diff
@@ -367,10 +412,24 @@ export class SSRService implements SSRServiceLike {
...
           projectSlug: ctx.projectSlug,
         });
-        return buildRedirectResult(outcome, slug);
+        return buildRedirectResult({
+          ...outcome,
+          ...mergeDataResponseMetadata([
+            inheritedResponseMetadata,
+            outcome,
+            requestLocalMetadata,
+          ]),
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/server/services/rendering/ssr.service.ts:421]: Fixed in `06c175580`. Buffered redirect and not-found handling now merges inherited metadata, then request-local loader metadata, then the later control metadata. The red-green regression covers both branches, duplicate header precedence, and loader-before-control cookie order, matching the existing late-streaming behavior.

File: src/data/schemas/data.schema.ts [hunk 20]
```diff
@@ -18,13 +28,43 @@ export const getRedirectSchema = defineSchema((v) =>
...
 );
 
+export const getResponseCookieSchema = defineSchema((v) =>
+  v.object({
+    name: v.string(),
+    value: v.string(),
+    domain: v.string().optional(),
+    path: v.string().optional(),
+    expires: v.string().optional(),
+    maxAge: v.number().optional(),
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/schemas/data.schema.ts:58]: Fixed in `06c175580`. `ResponseCookieSchema` now refines each parsed cookie through `normalizeDataResponseMetadata`, the same runtime validator used by execution and serialization. The regression covers fractional `maxAge`, invalid expiry, `__Secure-` and `__Host-` constraints, `SameSite=None`, and a valid host cookie.

**Comment by kojiwakayama** [comment]
Exact head `06c175580def0ce472ac67fa4f5a5bb70fef37e1` addresses both latest review findings. RED/GREEN coverage now proves buffered redirect and not-found controls override loader headers and follow loader cookies, while standalone response-cookie validation enforces the runtime constraints. Deno focused: 2 suites, 52 steps. Bun focused: 5 tests. Affected sweep: 130 suites, 2,196 steps. Formatting, lint, typecheck, test-typecheck baseline, generated API reference check, and diff check are green. @codex review

File: src/data/response-metadata.ts [hunk 21]
```diff
@@ -0,0 +1,385 @@
...
+  metadata: readonly DataResponseMetadata[],
+): DataResponseMetadata {
+  const headers = new Map<string, string>();
+  const cookies: ResponseCookie[] = [];
+  for (const value of metadata) {
+    const normalized = normalizeDataResponseMetadata(value);
+    for (const [name, headerValue] of objectEntries(normalized.headers ?? {})) {
+      headers.set(name, headerValue);
+    }
+    cookies.push(...(normalized.cookies ?? []));
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/response-metadata.ts:306]: Fixed in `be0f97339`. `mergeDataResponseMetadata` now re-normalizes the final merged object, so the same 64-header and 64-cookie limits apply across all loaders after duplicate-header precedence is resolved. RED coverage combined two individually valid loader metadata objects into 65 distinct headers and 65 cookies; both previously passed the merge and now fail at the data aggregation boundary with the existing actionable limit errors. The broader data/rendering/SSR sweep passes 49 suites and 807 steps.

**Comment by kojiwakayama** [comment]
Exact head `be0f97339` addresses the aggregate metadata-limit finding.

RED proved that two individually valid loader results could merge into 65 distinct headers or 65 cookies without failing. GREEN revalidates the final merged object after header override semantics, so aggregate limits now fail at the data merge boundary instead of during HTTP response construction.

Verification:
- focused response metadata suite: 7 steps passed
- data/rendering/SSR sweep: 49 suites, 807 steps passed
- targeted `deno check`, lint, and format: passed
- `git diff --check`: passed

The review thread is replied to and resolved. Replacement CI is running.

@codex review

File: src/rendering/renderer.ts [hunk 22]
```diff
@@ -847,6 +849,23 @@ export class Renderer {
       throw error;
     }
 
+    if (isFollower && cachedData.cookies?.length) {
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/renderer.ts:884]: Fixed in exact head 6583b036e. RED reproduced two concurrent callers inheriting the leader cookie when the shared render rejected. GREEN now detects cookie metadata on a shared failure and rerenders the follower outside singleflight with its own request. The test verifies distinct leader and follower cookies, two renders, and no cache write.

File: src/rendering/orchestrator/render-result-assembly.ts [hunk 23]
```diff
@@ -49,10 +52,12 @@ export function assembleRenderResult(options: AssembleRenderResultOptions): Rend
...
     ...(pageModule ? { pageModule } : {}),
   };
 
-  if (options.shouldCache && !options.skipCachePersist) {
+  if (options.shouldCache && !options.skipCachePersist && !options.cookies?.length) {
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/orchestrator/render-result-assembly.ts:60]: Fixed in exact head 6583b036e. RED showed CacheCoordinator.persistResult removed custom response headers before cloning the payload. GREEN copies headers into the lifecycle cache payload, and the regression verifies both stored and hydrated cache results retain the header.

**Comment by kojiwakayama** [comment]
Exact head 6583b036e is rebased onto current main and addresses both latest findings with RED-GREEN regressions.

- Cookie-bearing shared render failures now rerender followers outside singleflight, preserving request-local cookies for redirect, not-found, and post-data error paths.
- Lifecycle CacheCoordinator payloads now retain custom response headers through storage and hydration.
- Focused renderer and cache tests: 76 steps passed.
- Broader rendering, cache, orchestrator, and SSR sweep: 46 suites, 678 steps passed.
- Targeted check, lint, format, and git diff checks passed.
- Both review threads are replied to and resolved.

Replacement CI is running. @codex review

File: src/rendering/renderer.ts [hunk 24]
```diff
@@ -823,6 +826,25 @@ export class Renderer {
         )
         : await runRender();
     } catch (error) {
+      if (
+        isFollower && error instanceof Error &&
+        getAttachedDataResponseMetadata(error).cookies?.length
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/renderer.ts]: Fixed in exact head dad9e8013. RED reproduced concurrent leader/follower renders throwing raw cookie-bearing redirect controls and showed the follower receiving the leader cookie. GREEN resolves control metadata in the singleflight catch path and rerenders only cookie-bearing followers uncached, so each request receives its own cookie. The focused renderer suite passes 52 steps; the renderer/cache/SSR/orchestrator sweep passes 41 modules and 547 steps; typecheck, lint, format, and diff checks pass.

**Comment by kojiwakayama** [comment]
Exact-head update for `dad9e8013`: addressed the raw cookie-bearing SSR control replay finding with a RED/GREEN concurrent regression. Local verification is green: focused renderer 52 steps; renderer/cache/SSR/orchestrator sweep 41 modules, 547 steps; typecheck, lint, format, and diff checks. The review thread is resolved. @codex review

File: src/rendering/renderer.ts [hunk 25]
```diff
@@ -823,6 +827,29 @@ export class Renderer {
         )
         : await runRender();
     } catch (error) {
+      const attachedCookies = error instanceof Error
+        ? getAttachedDataResponseMetadata(error).cookies
+        : undefined;
+      const controlCookies = resolveSSRControlOutcome(error)?.cookies;
+      if (
+        isFollower &&
+        ((attachedCookies?.length ?? 0) > 0 || (controlCookies?.length ?? 0) > 0)
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/renderer.ts:843]: Fixed in 0b6af8642. The singleflight follower path now unwraps request-local metadata carriers before inspecting SSR controls, while retaining the wrapper fallback. Added a RED/GREEN concurrency regression that wraps a cookie-bearing redirect with header-only attached metadata and verifies leader/follower cookie isolation.

File: src/data/types.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/types.ts]: Fixed in 0b6af8642. PageWithData.getStaticData again accepts the legacy DataResult<T> annotation at the public boundary. Runtime validation still normalizes it as StaticDataResult and rejects headers/cookies before caching. Added the compatibility type regression, retained runtime rejection coverage, regenerated API references, and ran deno task lint:ci successfully.

**Comment by kojiwakayama** [comment]
Exact head 0b6af8642 addresses both latest findings with RED/GREEN regressions. Wrapped cookie-bearing controls now force request-local follower rerenders, legacy DataResult annotations remain assignable to getStaticData, and runtime static metadata rejection remains enforced. Focused renderer and data tests, compile checks, generated references, and the complete deno task lint:ci gate are green. Both threads are replied to and resolved. @codex review

File: src/data/schemas/data.schema.ts [hunk 26]
```diff
@@ -18,13 +37,43 @@ export const getRedirectSchema = defineSchema((v) =>
...
+);
+
 /** Result returned from data fetching functions */
 export const getDataResultSchema = defineSchema((v) =>
   v.object({
     props: v.unknown().optional(),
     redirect: getRedirectSchema().optional(),
     notFound: v.boolean().optional(),
-    revalidate: v.union([v.number(), v.literal(false)]).optional(),
+    revalidate: v.union([v.number().nonnegative(), v.literal(false)]).optional(),
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/data/schemas/data.schema.ts]: Fixed in 61c336ccd. The exported schemas now apply redirect and not-found control precedence before validating active revalidate values. RED reproduced rejection of redirect plus negative revalidate and notFound plus string revalidate; GREEN covers both controls while retaining rejection for negative props revalidate. Targeted schema/type tests and the complete lint:ci gate pass.

**Comment by kojiwakayama** [comment]
Review target: 61c336ccd9ba744a261de18888b84234b461349f. All known review findings are fixed, replied to, and resolved. Complete local lint:ci and focused RED-GREEN schema validation pass. @codex review

File: src/rendering/orchestrator/pipeline.ts [hunk 27]
```diff
@@ -635,23 +653,47 @@ export class RenderPipeline {
...
+    const responseMetadata = mergeDataResponseMetadata(
+      [
+        ...dataResults.filter(({ type }) => type === "layout"),
+        ...dataResults.filter(({ type }) => type === "page"),
+      ]
+        .flatMap(({ result }) => result ? [result] : []),
+    );
+
+    const failedResult = dataResults.find(({ error }) => error);
+    if (failedResult?.error) {
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/orchestrator/pipeline.ts]: Fixed in f6b93e5f1. Metadata is still pre-merged from all successful data results, but errors are now handled inside the original ordered outcome loop. RED reproduced a page notFound followed by a layout failure returning the layout 500; GREEN preserves the earlier page control and its response metadata. The focused 44-step pipeline suite and complete lint:ci pass.

**Comment by kojiwakayama** [comment]
Review target: f6b93e5f1. The exact-head RED-GREEN fix restores page-first data control precedence over later loader failures while retaining pre-merged successful response metadata. All known findings are replied to and resolved. Focused pipeline suite: 44 steps green. Complete local lint:ci green. @codex review

File: src/rendering/orchestrator/pipeline.ts [hunk 28]
```diff
@@ -799,6 +845,8 @@ export class RenderPipeline {
                             ? dataResolution.pageProps
                             : undefined;
                           layoutDataMap = dataResolution.layoutProps;
+                          responseHeaders = dataResolution.headers;
+                          responseCookies = dataResolution.cookies;
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/rendering/orchestrator/pipeline.ts:849]: Fixed in exact head 7f0422f97. RED: the new script-page pipeline regression showed getServerData headers and cookies were absent from the scriptResult early return. GREEN: the pipeline now merges validated script and data-response metadata before returning; the full pipeline behavior suite passes 45 steps, formatting and git diff checks pass, and deno task lint:ci is green.

**Comment by kojiwakayama** [comment]
Review target: exact head `7f0422f97`. The script-page metadata finding is fixed with RED-GREEN coverage, the review reply is posted, and its thread is resolved. Full local `deno task lint:ci` is green. Replacement CI is running. @codex review
